doc/designs/functions-for-explicitly-fetched-signature-algorithms.md

```diff
@@ -0,0 +1,181 @@
+/* Oneshot functions */
+int EVP_PKEY_verify_primitive(EVP_PKEY_CTX *pctx,
+                              const unsigned char *sig, size_t siglen,
+                              const unsigned char *tbs, size_t tbslen);
+int EVP_PKEY_verify_recover_primitive(EVP_PKEY_CTX *pctx,
+                                      unsigned char *rout, size_t *routlen,
+                                      size_t routsize,
+                                      cnnst unsigned char *sig, size_t siglen);
+```
+
+#### Streaming
+
+Do note that there's no proposed streaming definition for verify recover
+functionality.
+
+``` C
+/* Initializers */
+int EVP_PKEY_verify_data_init(EVP_PKEY_CTX *pctx, EVP_SIGNATURE *algo,
+                              const OSSL_PARAM params[]);
+/* Signature setter */
+int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,
+                               unsigned char *sig, size_t siglen,
+                               size_t sigsize);
+/* Update and final */
+int EVP_PKEY_verify_data_update(EVP_PKEY_CTX *ctx,
+                                const unsigned char *in, size_t inlen);
+int EVP_PKEY_verify_data_final(EVP_PKEY_CTX *ctx);
+```
+
+
+Requirements on the providers
+-----------------------------
+
+Because it's not immediately obvious from a composite algorithm name what
+key type it requires / supports, at least in code, allowing the use of an
+explicitly fetched implementation of a composite algorithm requires that
+providers cooperate by declaring what key type is required / supported by
+each algorithm.
+
+For non-composite operation algorithms (like "RSA"), this is not necessary,
+see the fallback strategies below.
+
+This is to be implemented through an added provider function that would work
+like keymgmt's `query_operation_name` function, but would return a NULL
+terminated array of key type name instead:
+
+``` C
+# define OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE         26
+OSSL_CORE_MAKE_FUNC(const char **, signature_query_key_type, (void))
+```
+
+Furthermore, the public API above requires added provider functionality:
+
+``` C
+# define OSSL_FUNC_SIGNATURE_SIGN_UPDATE            26
+# define OSSL_FUNC_SIGNATURE_SIGN_FINAL             27
+OSSL_CORE_MAKE_FUNC(int, signature_sign_update, (void *ctx,
+                                                 const unsigned char *in,
+                                                 size_t inlen))
+OSSL_CORE_MAKE_FUNC(int, signature_sign_final, (void *ctx,  unsigned char *sig,
+                                                size_t *siglen, size_t sigsize))
+
+# define OSSL_FUNC_SIGNATURE_VERIFY_UPDATE          28
+# define OSSL_FUNC_SIGNATURE_VERIFY_FINAL           29
+OSSL_CORE_MAKE_FUNC(int, signature_verify_update,
+                    (void *ctx, const unsigned char *in, size_t inlen))
+/*
+ * signature_verify_final requires that the signature to be verified against
+ * is specified via an OSSL_PARAM.
+ */
+OSSL_CORE_MAKE_FUNC(int, signature_verify_final, (void *ctx))
```

Comment:
I assume that we need to reflect the EVP api calls in the providers interface as well?

Reply:
No, please no. As a provider author, I *protest* this proliferation of a gazillion functions that do the same essential thing

Reply:
I do not understand how we can implement the proposed EVP calls in terms of the existing provider calls.

Reply:
The only additional calls I can see making sense would be `signature_sign_data_init` and `signature_verify_data_init`.  If that was all you meant, then sure.

But for example, I don't see why we'd need to have two different oneshot functions.

Reply:
I was actually thinking about not adding additional _init functions because the existing ones could cover both sign_data and sign_primitive. But then you of course need two different oneshot functions because otherwise you cannot distinguish whether you want the primitive operation or not.

Reply:
If we have a different init operation for the primitive and data calls then the oneshot op can be the same, that's true.

Reply:
Actually...  I don't see much of a need to have different inits, even.  `primitive` vs `data` is really semantic sugar for the application, 'cause in the end, what differentiates the "primitive" vs "data" functionality is that the former requires calling `EVP_PKEY_CTX_set_signature_md()` while the latter doesn't (for "data", it should even be prohibited - not settable - since the "data" algorithms are expected to be composite algorithms, i.e. hash function implied).

In terms of oneshot, the difference between "primitive" and "data" is really that the former might error with some sort of error indicating that the input (tbs) is too large.

The flip side with having alternative function indexes (one for "primitive" and one for "data") is the possible implication that you should be able to have both.  That's a source of unnecessary error and confusion for provider authors.

Reply:
In the end, there are a number of possible indicators to see if an `EVP_SIGNATURE` can be considered "primitive", which can be checked in libcrypto:

- The presence or not of `OSSL_FUNC_SIGNATURE_SIGN_UPDATE` and `OSSL_FUNC_SIGNATURE_VERIFY_UPDATE`.
- The presence or not of `OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE`
- If there is a `OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE`, does its response match the key type of the given (through `EVP_PKEY_CTX`) `EVP_PKEY`?

Reply:
I am afraid this will be very confusing to provider authors.

Reply:
With my provider author hat on, I disagree.  I find the multitude of diverse functions to implement a much bigger pile of confusion...  and error prone outside of the provider's control (hard question: did libcrypto call the right variant of, say, update, for this use case?)